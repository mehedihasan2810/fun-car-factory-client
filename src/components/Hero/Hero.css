.hero-container,
.hero-scene {
  position: relative;
  isolation: isolate;
  width: 100vw;
  height: 100vh;
}

.hero-container::before {
  content: "";
  position: absolute;
  top: 50%;
  left: 50%;
  z-index: -1;
  width: 400px;
  height: 400px;
  transform: translate(-50%, -50%);
  background-color: var(--primary-sky-400);
  border-radius: 50%;
  filter: blur(180px);
  opacity: 0.9;
}

.hero-layer-contaner {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 2;
  width: 100%;
  height: 100%;
  display: grid;
  place-items: center;
}

.mobile-hero-img {
  width: 100%;
  height: auto;
  display: none;
}

.hero-info {
  text-align: center;
}

.hero-info h1 {
  font-size: clamp(2rem, 6vw, 8rem);
  font-weight: 900;
  line-height: 1.3;
  font-family: "Brice Bold";
  color: var(--deep-white-100);
  text-shadow: 4px 4px 0px #fff;
}

.hero-highlight-title {
  display: inline-block;
  position: relative;
  isolation: isolate;
}

.hero-highlight-title svg {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  transform: translateY(-200%);
}

.hero-path {
  stroke-dasharray: 1528;
  stroke-dashoffset: 1528;
}

.hero-info a {
  display: inline-block;
  margin-block-start: 1rem;
}

/* -----------------------------
 hero image slider styles starts
 ------------------------------- */

.hero-scene {
  perspective: 400px;
  overflow: hidden;
}
.hero-slider,
.hero-slider-2 {
  position: absolute;
  top: 50%;
  width: 100vw;
  height: 100vh;
  transform-style: preserve-3d;
  overflow: hidden;
}

.hero-slider {
  transform: rotateY(-60deg) translate(0, -50%);
  transform-origin: top right;
}

.hero-slider-2 {
  transform: rotateY(60deg) translate(0, -50%);
  transform-origin: top left;
}

.hero-slider div,
.hero-slider-2 div {
  position: absolute;
  width: calc(100vh + 100px);
  height: 100vh;

  padding: 1rem;
}

.hero-slider div {
  left: calc(-100vh - 100px);
  perspective: 600px;
}
.hero-slider-2 div {
  right: calc(-100vh - 100px);
}

.hero-slider img {
  transform-style: preserve-3d;
  transform: rotateY(180deg);
}

.hero-slider img,
.hero-slider-2 img {
  width: 100%;
  height: 100%;
  object-fit: cover;
  border-radius: 1rem;
  opacity: 0.6;
}
/* -----------------------------
 hero image slider styles ends
 ------------------------------- */

@media (max-width: 799px) {
  .hero-info h1 {
    text-shadow: 2px 2px 0px #fff;
  }
  .mobile-hero-img {
    display: block;
  }
  .hero-scene {
    display: none;
  }
  .hero-container,
  .hero-scene {
    width: unset;
    height: unset;
  }
  .hero-container {
    margin-block: 8rem;
  }
  .hero-layer-contaner {
    all: unset;
  }
  .hero-container::before {
    width: 80vw;
    height: 80vw;
  }
}
