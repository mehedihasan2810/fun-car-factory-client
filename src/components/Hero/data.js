export const heroSlider1 = [
  {
    imgUrl:
      "https://images.unsplash.com/photo-1572635196184-84e35138cf62?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2080&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1594787317357-dcda50fd1d78?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1610018817073-fe06110e6693?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1565438525268-adf2e49737e4?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1931&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1616850508698-8cb0c576d5ae?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2038&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1559816933-0b5b4de2bbcd?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1974&q=80",
  },
];

export const heroSlider2 = [
  {
    imgUrl:
      "https://images.unsplash.com/photo-1572635196184-84e35138cf62?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2080&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1594787317357-dcda50fd1d78?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1610018817073-fe06110e6693?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1565438525268-adf2e49737e4?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1931&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1616850508698-8cb0c576d5ae?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2038&q=80",
  },
  {
    imgUrl:
      "https://images.unsplash.com/photo-1559816933-0b5b4de2bbcd?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1974&q=80",
  },
];
