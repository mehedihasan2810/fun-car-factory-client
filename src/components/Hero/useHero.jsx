import { gsap } from "gsap";
import { ScrollTrigger } from "gsap/ScrollTrigger";
import { useLayoutEffect, useRef } from "react";

gsap.registerPlugin(ScrollTrigger);

const useHero = () => {
  // Ref to the hero section container
  const heroConRef = useRef();

  useLayoutEffect(() => {
    // Dimensions and time variables for slider animation
    const boxWidth = window.innerHeight + 100,
      totalWidth = boxWidth * 6,
      time = 70;

    // Slider items for both directions
    const sliederItems = document.querySelectorAll(".hero-slider > div"),
      dirFromLeft = "+=" + totalWidth;
    const mod = gsap.utils.wrap(0, totalWidth);
    gsap.set(sliederItems, {
      x: (i) => i * boxWidth,
    });

    const sliederItems2 = document.querySelectorAll(".hero-slider-2 > div");
    const dirFromRight = "-=" + totalWidth;
    const mod2 = gsap.utils.wrap(0, -totalWidth);
    gsap.set(sliederItems2, {
      x: (i) => -1 * i * boxWidth,
    });

    // Media query to determine if it's a desktop
    const matchMedia = gsap.matchMedia();

    matchMedia.add(
      {
        isDesktop: "(min-width: 800px)",
      },
      () => {
        // Timeline for left to right slider animation
        const tl1 = gsap.timeline().to(sliederItems, {
          x: dirFromLeft,
          modifiers: {
            x: (x) => mod(parseFloat(x)) + "px",
          },
          duration: time,
          ease: "none",
          repeat: -1,
        });
        // slider right to left ends

        // Timeline for right to left slider animation
        const tl2 = gsap.timeline().to(sliederItems2, {
          x: dirFromRight,
          modifiers: {
            x: (x) => mod2(parseFloat(x)) + "px",
          },
          duration: time,
          ease: "none",
          repeat: -1,
        });
        // slider left to right ends

        // Pause slider animation when hero section is not in the viewport
        ScrollTrigger.create({
          trigger: heroConRef.current,
          start: "top bottom",
          end: "bottom top",
          onToggle: (s) => {
            if (!s.isActive) {
              tl1.pause();
              tl2.pause();
            } else {
              tl1.play();
              tl2.play();
            }
          },
        });
      },
      heroConRef.current
    );

    return () => {
      matchMedia.revert();
    };
  }, []);

  // Animation for the SVG path
  useLayoutEffect(() => {
    const path = document.querySelector(".hero-path");
    gsap.to(path, {
      strokeDashoffset: 0,
      duration: 1.5,
      ease: "power1",
    });
  }, []);

  return {
    heroConRef,
  };
};

export default useHero;
