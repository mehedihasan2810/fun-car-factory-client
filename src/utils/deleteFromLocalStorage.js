const deleteFromLocalStorage = (key, id) => {
  if (localStorage.getItem(key)) {
    const items = JSON.parse(localStorage.getItem(key));
    const filteredItems = items.filter((item) => id !== item);
    localStorage.setItem(key, JSON.stringify(filteredItems));
  }
};

export default deleteFromLocalStorage;
