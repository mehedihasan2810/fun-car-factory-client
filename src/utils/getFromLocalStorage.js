const getFromLocalStorage = (key) => {
  return localStorage.getItem(key);
};

export default getFromLocalStorage;
