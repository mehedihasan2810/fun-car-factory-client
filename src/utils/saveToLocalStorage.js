const saveToLocalStorage = (id) => {
  if (localStorage.getItem("cart")) {
    const items = JSON.parse(localStorage.getItem("cart"));
    if (items.includes(id)) {
     throw new Error('Item already exists!')
    } else {
      items.push(id);
      localStorage.setItem("cart", JSON.stringify(items));
    }
  } else {
    localStorage.setItem("cart", JSON.stringify([id]));
  }
};

export default saveToLocalStorage;
