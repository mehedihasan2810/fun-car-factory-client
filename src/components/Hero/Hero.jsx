import { Link } from "react-router-dom";
import "./Hero.css";
import useHero from "./useHero";
import { heroSlider1, heroSlider2 } from "./data";
const Hero = () => {
  // Custom hook to handle hero section functionality
  const { heroConRef } = useHero();

  return (
    // Hero section container
    <section ref={heroConRef} className="hero-container">
      <div className="hero-layer-contaner">
        {/* Mobile hero image */}
        <img className="mobile-hero-img" src="/assets/banner-photo.png"></img>

        {/* Hero information */}
        <div className="hero-info">
          <h1>
            Get The {/* Highlighted title with SVG underline */}
            <div className="hero-highlight-title">
              Best Toys
              <svg
                viewBox="0 0 543 11"
                fill="none"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  className="hero-path"
                  d="M0.5 1.5H250L0.5 9.5H250H542.5L299.5 1.5H542.5"
                  stroke="#082f49"
                  strokeWidth="5"
                />
              </svg>
            </div>{" "}
            <br /> For Your Kids
          </h1>

          {/* Button to navigate to all toys */}
          <Link to="/all-toys" className="btn-primary">
            Get Now
          </Link>
        </div>
      </div>

      {/* Hero scene with slider */}
      <div className="hero-scene">
        <div className="hero-slider">
          {heroSlider1.map((item, i) => (
            <div key={i}>
              <img src={item.imgUrl} alt="" aria-hidden />
            </div>
          ))}
        </div>

        {/* Second hero slider */}
        <div className="hero-slider-2">
          {heroSlider2.map((item, i) => (
            <div key={i}>
              <img src={item.imgUrl} alt="" aria-hidden />
            </div>
          ))}
        </div>
      </div>
    </section>
  );
};

export default Hero;
