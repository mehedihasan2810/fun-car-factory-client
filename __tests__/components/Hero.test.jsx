import { customRender, describe, expect, screen, test } from "../utils/utils";
import Hero from "../../src/components/Hero/Hero";

describe("<Hero/>", () => {
  test("should render successfully and have one heading and button", () => {
    // ARRANGE
    customRender(<Hero />);

    // ACT
    const title = screen.getByRole("heading");
    const button = screen.getByText("Get Now");

    // ASSERT
    expect(title).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });
  test("should match snapshot", () => {
    // ARRANGE
    const { asFragment } = customRender(<Hero />);

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });
});
