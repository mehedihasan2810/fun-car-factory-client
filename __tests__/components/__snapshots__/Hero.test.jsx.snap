// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Hero/> > should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="hero-container"
  >
    <div
      class="hero-layer-contaner"
    >
      <img
        class="mobile-hero-img"
        src="/assets/banner-photo.png"
      />
      <div
        class="hero-info"
      >
        <h1>
          Get The 
          <div
            class="hero-highlight-title"
          >
            Best Toys
            <svg
              fill="none"
              viewBox="0 0 543 11"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="hero-path"
                d="M0.5 1.5H250L0.5 9.5H250H542.5L299.5 1.5H542.5"
                stroke="#082f49"
                stroke-width="5"
              />
            </svg>
          </div>
           
          <br />
           For Your Kids
        </h1>
        <a
          class="btn-primary"
          href="/all-toys"
        >
          Get Now
        </a>
      </div>
    </div>
    <div
      class="hero-scene"
    >
      <div
        class="hero-slider"
      >
        <div
          style="transform: translate(0, 0);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1572635196184-84e35138cf62?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2080&q=80"
          />
        </div>
        <div
          style="transform: translate(868px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1594787317357-dcda50fd1d78?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80"
          />
        </div>
        <div
          style="transform: translate(1736px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1610018817073-fe06110e6693?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80"
          />
        </div>
        <div
          style="transform: translate(2604px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1565438525268-adf2e49737e4?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1931&q=80"
          />
        </div>
        <div
          style="transform: translate(3472px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1616850508698-8cb0c576d5ae?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2038&q=80"
          />
        </div>
        <div
          style="transform: translate(4340px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1559816933-0b5b4de2bbcd?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1974&q=80"
          />
        </div>
      </div>
      <div
        class="hero-slider-2"
      >
        <div
          style="transform: translate(0, 0);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1572635196184-84e35138cf62?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2080&q=80"
          />
        </div>
        <div
          style="transform: translate(-868px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1594787317357-dcda50fd1d78?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80"
          />
        </div>
        <div
          style="transform: translate(-1736px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1610018817073-fe06110e6693?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2070&q=80"
          />
        </div>
        <div
          style="transform: translate(-2604px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1565438525268-adf2e49737e4?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1931&q=80"
          />
        </div>
        <div
          style="transform: translate(-3472px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1616850508698-8cb0c576d5ae?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=2038&q=80"
          />
        </div>
        <div
          style="transform: translate(-4340px, 0px);"
        >
          <img
            alt=""
            aria-hidden="true"
            src="https://images.unsplash.com/photo-1559816933-0b5b4de2bbcd?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&w=1974&q=80"
          />
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
